src/users.py

```diff
@@ -0,0 +1,440 @@
+
+    if raw_nick:
+        if ident is not None or host is not None:
+            raise ValueError("ident and host need to be None if raw_nick is True")
+        nick, ident, host = parse_rawnick(nick)
+
+    potential = []
+    users = set(all_users)
+    if allow_bot:
+        users.add(Bot)
+
+    for user in users:
+        if nick is not None and user.nick != nick:
+            continue
+        if ident is not None and user.ident != ident:
+            continue
+        if host is not None and user.host != host:
+            continue
+        if realname is not None and user.realname != realname:
+            continue
+        if account is not None and user.account != account:
+            continue
+
+        if not potential or allow_multiple:
+            potential.append(user)
+        else:
+            raise ValueError("More than one user matches: " +
+                             _arg_msg.format(nick, ident, host, realname, account, allow_bot))
+
+    if not potential and not allow_none:
+        raise KeyError(_arg_msg.format(nick, ident, host, realname, account, allow_bot))
+
+    if allow_multiple:
+        return potential
+
+    if not potential: # allow_none
+        return None
+
+    return potential[0]
+
+def add(cli, *, nick, ident=None, host=None, realname=None, account=None, channels=None, raw_nick=False):
+    """Create a new user, add it to the user list and return it.
+
+    This function takes up to 6 keyword-only arguments (and no positional
+    arguments): nick, ident, host, realname, account and channels.
+    With the exception of the first one, any parameter can be omitted.
+    If a matching user already exists, a ValueError will be raised.
+
+    The raw_nick keyword argument may be set if the nick has not yet
+    been parsed. In that case, ident and host must both be None, and
+    nick must be in the form nick!ident@host.
+
+    """
+
+    if raw_nick:
+        if ident is not None or host is not None:
+            raise ValueError("ident and host need to be None if raw_nick is True")
+        nick, ident, host = parse_rawnick(nick)
+
+    if exists(nick, ident, host, realname, account, allow_multiple=True, allow_bot=True):
+        raise ValueError("User already exists: " + _arg_msg.format(nick, ident, host, realname, account, True))
+
+    if channels is None:
+        channels = {}
+    else:
+        channels = dict(channels)
+
+    cls = User
+    if predicate(nick):
+        cls = FakeUser
+
+    new = cls(cli, nick, ident, host, realname, account, channels)
+    all_users.add(new)
+    return new
+
+def exists(*args, allow_none=False, **kwargs):
+    """Return True if a matching user exists.
+
+    Positional and keyword arguments are the same as get(), with the
+    exception that allow_none may not be used (a RuntimeError will be
+    raised in that case).
+
+    """
+
+    if allow_none: # why would you even want to do that?
+        raise RuntimeError("Cannot use allow_none=True with exists()")
```

Comment:
You haven't really replied on IRC so I'll just copy it here for posterity. Sorry about the tone, I've been having a bad week (don't want to go into it), but I shouldn't be taking it out on you.

```
<woffle> Vgr: and since I didn't really delve into why on the review, APIs that can take any arbitrary arguments even though it only cares about some of them are terrible for developers to use. It kills any sort of editor help/completion you may have available, and it forces you to look in another completely separate spot to see how to actually use it, since the function signature itself no longer tells you
<woffle> If you're taking in *args or **kwargs, that better be because you actually do allow any arbitrary arguments, not because you're being lazy
<woffle> that exact same reason is why I killed your other change to events
<woffle> s/killed/re-did/
```